csrc/multidevice/executor.h

```diff
@@ -8,56 +8,75 @@
 #ifdef NVFUSER_DISTRIBUTED
 #pragma once
 
+#include <c10/core/DeviceType.h>
 #include <exceptions.h>
-#include <iter_visitor.h>
-#include <kernel_cache.h>
+#include <fusion.h>
+#include <fusion_segmenter.h>
 #include <multidevice/communication.h>
-#include <multidevice/pipeline_ir.h>
-#include <multidevice/runtime.h>
+#include <multidevice/communicator.h>
+#include <multidevice/multidevice.h>
 
 namespace nvfuser {
 
-// Runtime Executor for Pipelines
-// This class inherits from IterVisitor because the execution
-// is ordered by the traversal of the Pipeline seen as a DAG
-class PipelineExecutor : public IterVisitor {
+/*
+  The MultiDeviceExecutor executes a Fusion on a multi-device setting.
+  It is instantiated from a Fusion and a Communicator.
+*/
+class MultiDeviceExecutor {
  public:
-  explicit PipelineExecutor(MultiDeviceRuntime& runtime)
-      : IterVisitor(), runtime_(runtime) {}
+  MultiDeviceExecutor(std::unique_ptr<Fusion> fusion, Communicator& comm);
 
-  // Run the Pipelined Fusion with the given global inputs
+  // Run the fusion on several devices with the given global inputs
   std::vector<at::Tensor> runWithInput(const std::vector<c10::IValue>& inputs);
 
+  // Returns the Communicator
+  Communicator* comm() const {
+    return &comm_;
+  }
+
+  // Returns the Fusion
+  auto fusion() const {
```

Comment:
I'd call this `completeFusion` as it's often really confusing which fusion we are talking about.